### ISSUE TITLE
[feature] how to set lr_scheduler_type in xtuner conf file?

### ISSUE THREAD

**Issue by amulil** [InternLM/xtuner#211]
.

**Comment by LZHgrla** [comment]
Modify the `param_scheduler ` according to the [docs](https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/param_scheduler.md)
https://github.com/InternLM/xtuner/blob/0badead00dd43b5e25d1ea6edb0a52bd2626c856/xtuner/configs/internlm/internlm_7b/internlm_7b_qlora_oasst1_e3.py#L119-L125

**Comment by amulil** [comment]
> docs

thanks, I will see it.

**Comment by amulil** [comment]
@LZHgrla I see the train log, I find the lr is a constant in log.But I open the CosineAnnealingLR.

log info
```
param_scheduler = dict(
    T_max=20,
    by_epoch=True,
    convert_to_iter_based=True,
    eta_min=2e-05,
    type='mmengine.optim.CosineAnnealingLR')
Epoch(train)  [4][ 10/384]  lr: 2.0000e-05
Epoch(train) [19][ 10/384]  lr: 2.0000e-05
```
the lr is still 2e-05  at the end of train.Is it a bug?

**Comment by LZHgrla** [comment]
@amulil 
```
eta_min=2e-05,
```
What is the initial learning rate?

**Comment by amulil** [comment]
> @amulil
> 
> ```
> eta_min=2e-05,
> ```
> 
> What is the initial learning rate?

2e-04

**Comment by LZHgrla** [comment]
@amulil 
Would you please paste the full config? I will try to repro it!

**Comment by amulil** [comment]
> @amulil Would you please paste the full config? I will try to repro it! Please remember to mask the sensitive information, such as the path name.

@LZHgrla 
ok, the conf such as:
```
# Copyright (c) OpenMMLab. All rights reserved.
import torch
from torch.optim import AdamW
from datasets import load_dataset
from mmengine.dataset import DefaultSampler
from mmengine.hooks import (CheckpointHook, DistSamplerSeedHook, IterTimerHook,
                            LoggerHook, ParamSchedulerHook)
from mmengine.optim import AmpOptimWrapper, CosineAnnealingLR
from mmengine.visualization import Visualizer,WandbVisBackend, TensorboardVisBackend
from peft import LoraConfig
from transformers import (AutoModelForCausalLM, AutoTokenizer,
                          BitsAndBytesConfig)

from xtuner.dataset import ConcatDataset, process_hf_dataset
from xtuner.dataset.collate_fns import default_collate_fn
from xtuner.dataset.map_fns import (alpaca_map_fn, alpaca_zh_map_fn,
                                    oasst1_map_fn, medical_map_fn, template_map_fn_factory)
from xtuner.engine import DatasetInfoHook, EvaluateChatHook
from xtuner.model import SupervisedFinetune
from xtuner.utils import PROMPT_TEMPLATE, SYSTEM_TEMPLATE

#######################################################################
#                          PART 1  Settings                           #
#######################################################################
# Model
pretrained_model_name_or_path = './model/Yi-6B'

# Data
oasst1_path = 'timdettmers/openassistant-guanaco'

prompt_template = PROMPT_TEMPLATE.qwen_chat
max_length = 2048
pack_to_max_length = True

# Scheduler & Optimizer
batch_size = 4  # per_device
accumulative_counts = 4
dataloader_num_workers = 4
max_epochs = 20
optim_type = AdamW
lr = 2e-4
betas = (0.9, 0.999)
weight_decay = 0
max_norm = 1  # grad clip

# Evaluate the generation performance during the training
evaluation_freq = 500
SYSTEM = SYSTEM_TEMPLATE.alpaca
evaluation_inputs = [
    '请给我介绍五个上海的景点', 'Please tell me five scenic spots in Shanghai'
]

#######################################################################
#                      PART 2  Model & Tokenizer                      #
#######################################################################
tokenizer = dict(
    type=AutoTokenizer.from_pretrained,
    pretrained_model_name_or_path=pretrained_model_name_or_path,
    trust_remote_code=True,
    padding_side='right')

model = dict(
    type=SupervisedFinetune,
    llm=dict(
        type=AutoModelForCausalLM.from_pretrained,
        pretrained_model_name_or_path=pretrained_model_name_or_path,
        trust_remote_code=True,
        torch_dtype=torch.float16,
        quantization_config=dict(
            type=BitsAndBytesConfig,
            load_in_4bit=True,
            load_in_8bit=False,
            llm_int8_threshold=6.0,
            llm_int8_has_fp16_weight=False,
            bnb_4bit_compute_dtype=torch.float16,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_quant_type='nf4')),
    lora=dict(
        type=LoraConfig,
        r=16,
        lora_alpha=16,
        lora_dropout=0.05,
        target_modules=['gate_proj', 'down_proj', 'up_proj'],
        bias='none',
        task_type='CAUSAL_LM'))

#######################################################################
#                      PART 3  Dataset & Dataloader                   #
#######################################################################
oasst1 = dict(
    type=process_hf_dataset,
    dataset=dict(type=load_dataset, path=oasst1_path),
    tokenizer=tokenizer,
    max_length=max_length,
    dataset_map_fn=oasst1_map_fn,
    template_map_fn=dict(
        type=template_map_fn_factory, template=prompt_template),
    remove_unused_columns=True,
    shuffle_before_pack=True,
    pack_to_max_length=pack_to_max_length)

train_dataset = dict(
    type=ConcatDataset,
    datasets_cfg=dict(
        oasst1=oasst1)

train_dataloader = dict(
    batch_size=batch_size,
    num_workers=dataloader_num_workers,
    dataset=train_dataset,
    sampler=dict(type=DefaultSampler, shuffle=True),
    collate_fn=dict(type=default_collate_fn))

#######################################################################
#                    PART 4  Scheduler & Optimizer                    #
#######################################################################
# optimizer
optim_wrapper = dict(
    type=AmpOptimWrapper,
    optimizer=dict(
        type=optim_type, lr=lr, betas=betas, weight_decay=weight_decay),
    clip_grad=dict(max_norm=max_norm, error_if_nonfinite=False),
    accumulative_counts=accumulative_counts,
    loss_scale='dynamic',
    dtype='float16')

# learning policy
# More information: https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/param_scheduler.md  # noqa: E501
param_scheduler = dict(
    type=CosineAnnealingLR,
    eta_min=lr * 0.1,
    by_epoch=True,
    T_max=max_epochs,
    convert_to_iter_based=True)

# train, val, test setting
train_cfg = dict(by_epoch=True, max_epochs=max_epochs, val_interval=1)

#######################################################################
#                           PART 5  Runtime                           #
#######################################################################
# Log the dialogue periodically during the training process, optional
custom_hooks = [
    dict(type=DatasetInfoHook, tokenizer=tokenizer),
    dict(
        type=EvaluateChatHook,
        tokenizer=tokenizer,
        every_n_iters=evaluation_freq,
        evaluation_inputs=evaluation_inputs,
        system=SYSTEM,
        prompt_template=prompt_template)
]

# configure default hooks
default_hooks = dict(
    # record the time of every iteration.
    timer=dict(type=IterTimerHook),
    # print log every 100 iterations.
    logger=dict(type=LoggerHook, interval=10),
    # enable the parameter scheduler.
    param_scheduler=dict(type=ParamSchedulerHook),
    # save checkpoint per epoch.
    checkpoint=dict(type=CheckpointHook, interval=1),
    # set sampler seed in distributed evrionment.
    sampler_seed=dict(type=DistSamplerSeedHook),
)

# configure environment
env_cfg = dict(
    # whether to enable cudnn benchmark
    cudnn_benchmark=False,
    # set multi process parameters
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    # set distributed parameters
    dist_cfg=dict(backend='nccl'),
)

# set visualizer
visualizer = dict(
    type=Visualizer, 
    vis_backends=[dict(type=TensorboardVisBackend)]
)

# set log level
log_level = 'INFO'

# load from which checkpoint
load_from = None

# whether to resume training from the loaded checkpoint
resume = False

# Defaults to use random seed and disable `deterministic`
randomness = dict(seed=None, deterministic=False)
```

**Comment by LZHgrla** [comment]
@amulil 
I cannot repro it... 

Could you please paste the `param_scheduler` and `optim_wrapper` information in the log? 
You can fine them by `head 20231108_182256/20231108_182256.log -n 260`)

**Comment by amulil** [comment]
> head

@LZHgrla 
I forget a info.I use --resume.

**Comment by LZHgrla** [comment]
@amulil 
I tried using `resume` but did not reproduce the issue. The learning rate schedule works fine...
Does this learning rate error can be reproduced on your setting?

**Comment by amulil** [comment]
> @amulil I tried using `resume` but did not reproduce the issue. The learning rate schedule works fine... Does this learning rate error can be reproduced on your setting?

@LZHgrla I reproduce this problem by:
```python3
CUDA_VISIBLE_DEVICES=7 xtuner train yi_6b_qlora_alpaca_enzh_e3 # set max_epochs = 1 then
CUDA_VISIBLE_DEVICES=7 xtuner train yi_6b_qlora_alpaca_enzh_e3  --resume ./xtuner/work_dirs/yi_6b_qlora_alpaca_enzh_e3/epoch_1.pth # change max_epochs = 2, now you can see the `lr` is a constant.

```

**Comment by LZHgrla** [comment]
Yes, you are right and now I can repro this bug. 
I will consult with my teammates at MMEngine, please wait for the news!

> > @amulil I tried using `resume` but did not reproduce the issue. The learning rate schedule works fine... Does this learning rate error can be reproduced on your setting?
> 
> @LZHgrla I reproduce this problem by:
> 
> ```python
> CUDA_VISIBLE_DEVICES=7 xtuner train yi_6b_qlora_alpaca_enzh_e3 # set max_epochs = 1 then
> CUDA_VISIBLE_DEVICES=7 xtuner train yi_6b_qlora_alpaca_enzh_e3  --resume ./xtuner/work_dirs/yi_6b_qlora_alpaca_enzh_e3/epoch_1.pth # change max_epochs = 2, now you can see the `lr` is a constant.
> ```

**Comment by LZHgrla** [comment]
@amulil 
There is no solution. `lr_scheduler` will be loaded when resuming. If modifying `max_epoch` is indeed necessary, using a fixed `lr_scheduler` is the best choice.
